### ISSUE TITLE
VAE visulazation problem

### ISSUE THREAD

**Issue by yufeng9819** [nv-tlabs/LION#43]
Hey,

I visualize my VAE training result and find that it cannot performance as well as your results show in tensorboard.

My visualization results are as follows:

![image](https://user-images.githubusercontent.com/113972543/233891555-31d95b2b-5b55-4f36-95d3-5ad29d110ca9.png)

Results above are obtained from training 4000 epochs by train_prior_clip with batch_size 12 and lr 1e-4 (I do not change any parameters) on 8 V100.

I want to check  whether the VAE model is trained enough and then I visualize training results at around 5000 epochs. 

![image](https://user-images.githubusercontent.com/113972543/233892355-311c6149-d37e-414b-bac0-6684986b8a63.png)

But the results did not get better.

I want to know the reason why I cannot get a reasonable results after 20 days training. (Do I need to train further, but in general, this is an unacceptable length of training)

Looking forward to your reply. @ZENGXH

**Comment by yufeng9819** [comment]
I find the reason why training results are  bad.

I further visualize the best_eval.pth and then find that the model is well trained in early epochs.

The results are as follows:
![截屏2023-04-24 14 10 47](https://user-images.githubusercontent.com/113972543/233913418-75bd1203-8f7b-4223-a279-37d4c048ada7.png)

**Comment by yufeng9819** [comment]
The best_val.pth is saved at the epoch 199.

In this training process, I do not manage to launch tensorboard, cause I find that at utils.utils.py, USE_TFB = int(os.environ.get('USE_TFB', 0)) (line 27), is always False. So it fail to 'init TFB'.

As your training setting, I only save checkpoints at epoch 1999, 3999,  best_eval.pth and snapshot. I also visualize the checkpoints at epoch 1999. The results are as follows: 
![image](https://user-images.githubusercontent.com/113972543/234163969-f78a07ee-459e-4f5b-ac2a-d61636b3156b.png)

The results  also seem to be bad.

Another problem is in trainer.eval_sample, it requires to load ref_val_%s.pt  (in fact, it load ref_val_all.pt for all cats of pointflow data). But in READEME.md, I only find PF2_val_all.pt. Is it the right reference information for evaluating all cats of pointflow datasets.

**Comment by yufeng9819** [comment]
I find that in training log, model calculate CM and EMD every 200 epochs. Then I manually visualize these result:
![CD](https://user-images.githubusercontent.com/113972543/234167969-33a1f233-31a8-4be0-938a-bccf810d39f9.png)
![EMD](https://user-images.githubusercontent.com/113972543/234167973-673e9d4c-7bb2-45fa-a5b1-0f0741bda9ce.png)
CD and EMD  increase in almost training process.

**Comment by yufeng9819** [comment]
Furthermore, I want to know the advantage of LION. 

To my knowledge,  one of advantages of latent diffusion model is to speed up training process. 
However, in LION, it consists of two latent code and their dimension are 128 and 8192 (=2048*3), which means LION cannot speed up the whole training process. Can I understand in this way?

Looking forward to your reply. @ZENGXH

**Comment by ZENGXH** [comment]
1. it's wired that your EMD and CD grow much faster; are you using `script/train_vae_all.sh`? 
2. the trend that EMD and CD continue increasing is expected, since the KL weight is increasing over 1/2 of the total epochs, and as the KL weight increase, the model keep smoothing / compressing the latent space (this make the learning of the diffusion model easier), and sacrificing the reconstruction. 
3. Re:  ref_val_all.pt VS  PF2_val_all.pt: the difference is that the PF2 is normalized with `norm_box`, so if you are using `script/train_vae_all.sh` they are the same for evaluation; Since `train_vae_all` use `norm_box` and it normalize of the val data 
4. `eval_sample` is used to evaluate the samples from prior: are you training the prior now? 
5. for E1999, it seems the better ckpt happens before that.. How long does it take to train 1 epoch? Is it possible that you could re-launch the job with (1) smaller snapshot period, like 200 (`viz.save_freq 200`); (2)reduce the vis freq (`viz.viz_freq -200`). And (3) trained for 1K epoch. If you are going to re-launch the job, I suggest you logging them to `comet_ml` (setting up the account is free), it will be much easier to monitor the progress. Or you can try turn on the Tensorboard/wandb (but I havn't use it for a long time so I am not sure whether they are still compatible or not). 
6. The advantage of LION is the increase of model expressivity&capacity. Latent diffusion model is not always (trained) faster than output space diffusion model: it depends on the goal of designing the model. For example, if you want to speed up the training / inference, you may want to make the latent space smaller. But in LION our goal is to have a high performance point cloud generator. Our latent space is designed in a way that the 1D latent can capture some global information about the shape, and such hierarchical latent space improve the performance especially when the model is trained on large scale & multi-class data.

**Comment by yufeng9819** [comment]
Re:
1. The results above are trained by script/train_vae_all.sh.
2. I train the VAE by cript/train_vae_all.sh and change the lr from 1e-4 to 1e-3 and batchsize to 24. Also I visualize the training result every day (for about 300 epcochs every day). The result are as follows:
    -  CD and EMD: The upward curve is steeper than the previous training process
![截屏2023-04-28 13 39 20](https://user-images.githubusercontent.com/113972543/235063565-3daf7735-1d05-4fbd-91f8-25c8956bf444.png)
    - epoch 633: 
![image](https://user-images.githubusercontent.com/113972543/235063814-44af81db-77c7-4994-9643-d9291cd0c437.png)
    - epoch 843: 
![image](https://user-images.githubusercontent.com/113972543/235064080-f740e328-6a5f-43de-85cb-3d9474d9516f.png)
    - epoch 1158:
![image](https://user-images.githubusercontent.com/113972543/235064362-42dc1498-279e-4207-84b6-e1705a43323e.png)
In this training process, the visualization results seem to be normal, since the model reconstruct great result while the latent gradually contains more and more noise. But the CD and MMD curve seem to be abnormal. @ZENGXH 

4. I got it, and I comment the eval_sample when I analysis VAE training results.

**Comment by ZENGXH** [comment]
* what's the difference between this MMD-CD plot verse the one [here](https://github.com/nv-tlabs/LION/issues/43#issuecomment-1521110462)? Are they from the same run? It seems the scale is pretty different?
* It seems the training is in a good shape anyway! Perhaps it just need a few hundreds of extra epoch to smooth out. In case you didn't change the snapshot freq (i.e., the ckpt is saved every 2k epoch), you can try to copy the snapshot.pt file manually (the snapshot.pt is saved every 30 min [here](https://github.com/nv-tlabs/LION/blob/13b1cf3b57ef0231c30637eb13572f628fb3cb96/trainers/base_trainer.py#LL278C14-L278C14)) 
* on shapenet, I tried 3k points, and I also try scene level models, it can still work to some extend (I just have some initial trial, though)

**Comment by yufeng9819** [comment]
1. The MMD-CD plot is the result of my second run.
2. I manually change the save freg to 200 by stop training and then resume the model to continue to train
![截屏2023-04-28 17 31 40](https://user-images.githubusercontent.com/113972543/235111469-1a2bfae9-1f00-49fb-a2b9-50d77c2308c4.png), and before that I manually save snapshot every time when I visualize VAE results by Tensorboard.
![截屏2023-04-28 17 33 01](https://user-images.githubusercontent.com/113972543/235111868-2858ec4c-c0c6-4ca3-ace2-79ef62e9170d.png). I hope I can get reasonable results in the end! 
3. Thanks for your kind respond! @ZENGXH
